### ISSUE TITLE
[405] Remote package tasks should have a retry - too nitpicky?

### ISSUE THREAD

**Issue by geerlingguy** [ansible/ansible-lint#456]
# Issue Type

- Feature request

# Ansible and Ansible Lint details

```
$ ansible --version
ansible 2.7.5
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/Users/jgeerling/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.15 (default, Jul 17 2018, 19:41:20) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]

$ ansible-lint --version
ansible-lint 4.0.0
```

- ansible installation method: pip
- ansible-lint installation method: pip

# Desired Behaviour

Using package modules (`apt`, `yum`, `dnf`, `package`, etc.) should not require the user to always add 3+ extra lines of configuration per use to resolve 405.

Rather, if it's truly important enough that we add retries that it should be a lint warning... maybe it should be the responsibility of ansible itself, and it could be a module parameter?

So I guess this is more of a 'question request'—since 405 hits pretty much every playbook, role, and tiny ansible script I've ever written, and would require me adding thousands of lines of the exact same thing to all those hundreds of instances:

```yaml
register: task_result
until: task_result is success
retries: 10
delay: 2
```

...might it not be better to make this the responsibility of Ansible itself? Otherwise, this is one of those warnings I think many might just drop. Seems extremely nitpicky to me (even when rated as a `LOW`).

# Actual Behaviour (Bug report only)

The task:

```
- name: Update apt cache (Debian).
  apt: update_cache=true cache_valid_time=86400
  when: ansible_os_family == 'Debian'
```

(or any task with apt, yum, dnf, package, etc.)

results in:

```
[405] Remote package tasks should have a retry
/Users/jgeerling/Dropbox/VMs/roles/geerlingguy.git/tasks/main.yml:2
Task/Handler: Ensure git is installed (RedHat).
```

**Comment by geerlingguy** [comment]
It also is a little more annoying to me, probably, because most of my roles support 4+ OSes, so many of them have 5-10 or more package-related tasks, and that racks up a _lot_ of notices!

For example, my MySQL role:

```
Starting import: task_id=348865, repository=geerlingguy/ansible-role-mysql

===== LINTING ROLE =====
yamllint OK.
ansible-lint Warnings:
./galaxy/downloads/tmpaWPbQE/tasks/setup-Archlinux.yml:2: [E405] Remote package tasks should have a retry
./galaxy/downloads/tmpaWPbQE/tasks/setup-Archlinux.yml:5: [E405] Remote package tasks should have a retry
./galaxy/downloads/tmpaWPbQE/tasks/setup-Debian.yml:6: [E405] Remote package tasks should have a retry
./galaxy/downloads/tmpaWPbQE/tasks/setup-Debian.yml:14: [E405] Remote package tasks should have a retry
./galaxy/downloads/tmpaWPbQE/tasks/setup-Debian.yml:17: [E405] Remote package tasks should have a retry
./galaxy/downloads/tmpaWPbQE/tasks/setup-RedHat.yml:2: [E405] Remote package tasks should have a retry
./galaxy/downloads/tmpaWPbQE/tasks/setup-RedHat.yml:9: [E405] Remote package tasks should have a retry
```

And it's not nearly as complex as some others I maintain.

**Comment by txdavec** [comment]
Just to +1 this and add another possible concern...
yum and dnf both seem to retry 10 times by default:

> retries Set the number of times any attempt to retrieve a file should retry before returning an error. Setting this to '0' makes yum try forever. Default is '10'.
https://linux.die.net/man/5/yum.conf
https://dnf.readthedocs.io/en/latest/conf_ref.html

You can check my math but if I start putting:
```yaml
register: task_result
until: task_result is success
retries: 10
delay: 2
```
on all my yum/dnf tasks am I not actually setting up 100 retries? This is fairly misleading to the end user who likely thinks they are getting something reasonable like 10 but are in fact getting each ansible retry multiplied by the package manager's default.
apt has a similar 'retries' setting that could possibly be in play depending on how the user has their apt.conf setup.

Given that different package managers are already aware of the _'Package operations are unreliable'_ issue and are in some cases already handling it I'm not sure it makes sense to treat every package manager equally and request this be added across the board.

**Comment by webknjaz** [comment]
This looks like a topic for proposal https://github.com/ansible/proposals/issues/new

**Comment by geerlingguy** [comment]
@webknjaz - Not sure about that—this is something that package managers sometimes handle on their own, sometimes don't, but as a general rule, I don't think we should try to enforce retries across the board on all the packaging-related modules.

As I put in the title, it seems too nitpicky. _Separately_, maybe retry logic could be added in Ansible by default for certain packaging-related modules _if those modules have a lot of flakiness_.

But in my experience, a failure of yum, apt, or dnf due to network issues is exceedingly rare... though most of my tests run in the US and EU :)

**Comment by awcrosby** [comment]
Looking at community feedback for this rule, we’ve had more comments with concerns than support. The concerns are all good points. A non-default rules folder (suggested here and historically in other issues) might be the best option, to move this rule from the default rules to “extra_rules”.

**Comment by awcrosby** [comment]
After more discussion we will drop rule 405 in the next release 4.1.0. If we find a greater need for a non-default rules folder, then that functionality can be discussed more and spec’d out.

**Comment by bbaassssiiee** [comment]
galaxy.ansible.com needs to update to ansible-lint >= 4.1.0 too.

**Comment by awcrosby** [comment]
galaxy.ansible.com has now been updated
